crates/node/src/cluster_state_refresher/mod.rs

```diff
@@ -0,0 +1,294 @@
+                        last_seen_alive: None,
+                    }))
+                } else {
+                    // we are not sure
+                    None
+                }
+            }
+            Some(ref seen) => {
+                let elapsed = seen.seen_at.elapsed();
+                if elapsed < suspect_threshold {
+                    Some(NodeState::Alive(AliveNode {
+                        generational_node_id: seen.generational_node_id,
+                        last_heartbeat_at: seen.seen_at,
+                    }))
+                } else if elapsed >= suspect_threshold && elapsed < dead_threshold {
+                    Some(NodeState::Suspect(SuspectNode {
+                        generational_node_id: seen.generational_node_id,
+                        last_attempt: seen.seen_at,
+                    }))
+                } else {
+                    Some(NodeState::Dead(DeadNode {
+                        last_seen_alive: Some(seen.seen_at),
+                    }))
+                }
+            }
+        }
+    }
+
+    async fn update_cluster_state(&self) -> Result<(), ShutdownError> {
+        let nodes = self.metadata.nodes_config_ref();
+
+        let mut cluster_state = ClusterState {
+            last_refreshed: Some(Instant::now().into_std()),
+            logs_metadata_version: self.metadata.logs_version(),
+            nodes_config_version: self.metadata.nodes_config_version(),
+            partition_table_version: self.metadata.partition_table_version(),
+            nodes: nodes
+                .iter()
+                .filter_map(|(node_id, _)| {
+                    let node_data = self.nodes.get(&node_id);
+                    let node_state = match node_data {
+                        None => None,
+                        Some(data) => self.node_state_for(data),
+                    };
+
+                    node_state.map(|state| (node_id, state))
+                })
+                .collect(),
+        };
+
+        cluster_state.nodes.insert(
+            self.metadata.my_node_id().as_plain(),
+            NodeState::Alive(AliveNode {
+                generational_node_id: self.metadata.my_node_id(),
+                last_heartbeat_at: MillisSinceEpoch::now(),
+            }),
+        );
+
+        trace!("cluster state: {:?}", cluster_state);
+
+        //todo: should we notify listener only if node liveness changes?
+        self.cluster_state_watch_tx
+            .send_replace(Arc::new(cluster_state));
+
+        Ok(())
+    }
+
+    #[instrument(level="debug", parent=None, skip_all, fields(
+        peer=msg.peer().to_string()
+    ))]
+    async fn on_pong(&mut self, mut msg: Incoming<NodePong>) -> Result<(), ShutdownError> {
+        msg.follow_from_sender();
+
+        trace!("Handling pong response");
+
+        let tracker = self.nodes.entry(msg.peer().as_plain()).or_default();
+        tracker.seen = Some(SeenState::new(msg.peer()));
+
+        Ok(())
+    }
```

Comment:
I'm really not sure if pong is needed. Nodes ping other nodes, and each node makes its own view based on the pings it has received.

Reply:
Yes, makes sense!